Automattic/woocommerce-payments
# Fix UPE methods checkbox selection

Fixes #4629 

#### Changes proposed in this Pull Request

The PR fixes the issue of UPE payment buttons getting unselected, in WC Pay settings, when multiple options are selected consecutively. 

##### Summary of the issue
The issue was happening due to a 1.5 second delay given for design reasons, to accomodate a possibility of a confirmation modal in certain cases. ( related to: paJDYF-2rk-p2 ). Multiple options clicked within the 1.5 second time gap would make only the last one checked, and would uncheck the previous ones. ( check video below )

**Video of before**

[Video of Before](https://user-images.githubusercontent.com/56378160/186328647-6e2e0175-75f3-44a1-96e9-33f7dbc41320.mov)

##### Summary of changes done

The function that managed [adding](https://github.com/Automattic/woocommerce-payments/blob/f5962d92e4b37ded818dcf1baf391a06cce943a0/client/payment-methods/index.js#L137) and [removing ](https://github.com/Automattic/woocommerce-payments/blob/develop/client/payment-methods/index.js#L144)payment methods from the `enabled_payment_methods_id` state array, took an outdated version of the complete state array, added the new payment method, and overwrote it fully. 

The change here makes sure that the associated actions take only the payment method ID that needs to be changed, and is either concatenated or removed from the most updated `enabled_payment_methods_id`  state array.  This makes sure that each check or uncheck action, affects only that checkbox ( payment method ) and not the rest. 

**After :**
![Screen Capture on 2022-11-25 at 11-11-08](https://user-images.githubusercontent.com/4162931/203910352-3e43ac14-dcfe-44fb-ae82-c03246d5719f.gif)


#### Testing instructions

* Go to WooCommerce -> Settings -> Payments
* Make sure you have UPE payment options enabled
* Find Payments accepted on checkout section
* Select at least two or three payment methods quickly
* Make sure that all the selected payment methods remain checked and do not roll back to unchecked
* Repeat the test with consecutively unchecking. You may not be able to uncheck multiple boxes quickly due to modals. 
* Test clicking okay or cancel in the removal confirmation modals. 



-------------------

- [x] Covered with tests (or have a good reason not to test in description ☝️) - Updated existing tests

**Post merge**
